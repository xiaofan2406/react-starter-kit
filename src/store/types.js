/* @flow */
declare type GreetSetMessageAction = {
  type: 'greet/SET_MESSAGE',
  payload: string,
};

declare type GreetSetTimesAction = {
  type: 'greet/SET_TIMES',
  payload: number,
};

declare type GreetResetAction = {
  type: 'greet/RESET',
};

declare type GreetState = {
  message: string,
  times: number,
};

declare type StoreState = {
  greet: GreetState,
};

declare type StoreAction =
  | GreetSetMessageAction
  | GreetSetTimesAction
  | GreetResetAction;
