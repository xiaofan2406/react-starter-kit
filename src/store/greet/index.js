import * as actions from './actions';
import * as selectors from './selectors';

export * from './actions';
export * from './selectors';
export { default as reducer } from './reducer';

export { actions, selectors };
