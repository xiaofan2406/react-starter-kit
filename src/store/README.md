# Redux Store

## Rules

Components should not be concerned with the the shape of the store, but they
should care about which _module_ it needs to import actions and selectors from.

Create an directory for each _module_ (separated depending on the business logic
of the application).

#### Inside a _module_

Each _module_ directory has at lease four files:

- **reducer.js**: Hosts the root reducer for this _module_, the reducer can be refactored into sub-reducer files.

- **actions.js**: Hosts all action creators for this _module_.

- **selector.js**: Hosts all selectors for this _module_. Each selector is aware of the _module_ key in the whole store. Therefore, there is no need for a global selector file.

- **index.js**: Exports all actions, selectors via named exports as well as namespace them udner `actions` and `selectors`. It also exports `reducer` for this _module_.

#### Inside the store

The root level `index.js` should export all _modules_ actions and selectors under the respective module's name.

**actions** and **selectors** shared between _modules_ should use relative import.

## Usage

```js
import { setMessage, getMessage } from 'store/greet';

import { actions as greetActions } from 'store/greet';
// greet.setMessage

import { selectors as greetSelectors } from 'store/greet';
// greetSelectors.getMessage

import { actions, selectors } from 'store';
// actions.greet.setMessage
// selectors.greet.getMessage
```
