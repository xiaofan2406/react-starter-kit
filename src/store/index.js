import { actions as greetActions, selectors as greetSelectors } from './greet';

export const actions = {
  greet: greetActions,
};

export const selectors = {
  greet: greetSelectors,
};

export { default as configureStore } from './configureStore';
