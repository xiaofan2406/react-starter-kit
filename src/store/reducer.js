import { combineReducers } from 'redux';

import { reducer as greetReducer } from './greet';

export default combineReducers({
  greet: greetReducer,
});
