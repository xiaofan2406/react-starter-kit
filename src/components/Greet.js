/* @flow */
import React from 'react';
import { connect } from 'react-redux';
import { css } from 'react-emotion';
import { theme, fontSizes, spacing, verticalScroll } from 'styles';
import { actions, selectors } from 'store/greet';

const cssGreet = css`
  margin: 0px auto;
  width: 480px;
  & > .controls {
    text-align: center;
    & input {
      border: none;
      border-bottom: 1px solid ${theme.borderColor};
      outline: none;
      width: 100px;
      text-align: center;
      &:focus {
        border-bottom: 1px solid ${theme.primaryColor};
      }
    }
    & > button {
      margin-left: 1em;
      outline: none;
      background: none;
      border: 1px solid ${theme.borderColor};
      &:hover {
        border-color: ${theme.primaryColor};
        color: ${theme.primaryColor};
      }
    }
  }
  & > .result {
    height: 280px;
    ${verticalScroll};
    margin-top: ${spacing.external}px;
    font-size: ${fontSizes.large}px;
    font-family: cursive;
    &:before {
      content: '"';
    }
    &:after {
      content: '"';
    }
  }
`;

type GreetProps = {
  message: string,
  setMessage: typeof actions.setMessage,
  times: number,
  setTimes: typeof actions.setTimes,
  reset: typeof actions.reset,
  greeting: string,
};

const Greet = ({
  message,
  setMessage,
  times,
  setTimes,
  reset,
  greeting,
}: GreetProps) => {
  const handleTimesOnChange = event => {
    setTimes(+event.currentTarget.value || 0);
  };

  const handleMessageOnChange = event => {
    setMessage(event.currentTarget.value);
  };

  return (
    <div className={cssGreet}>
      <div className="controls">
        <span>Say </span>
        <input
          data-testid="messageInput"
          type="text"
          onChange={handleMessageOnChange}
          value={message}
        />{' '}
        <input
          data-testid="timesInput"
          type="number"
          min={1}
          onChange={handleTimesOnChange}
          value={times}
        />{' '}
        times
        <button onClick={reset} type="button">
          Clear
        </button>
      </div>
      <div className="result">Hello {greeting}</div>
    </div>
  );
};

export const mapStateToProps = (state: StoreState) => ({
  message: selectors.getMessage(state),
  times: selectors.getTimes(state),
  greeting: selectors.getGreeting(state),
});

export { Greet as Component };

export default connect(
  mapStateToProps,
  {
    setMessage: actions.setMessage,
    setTimes: actions.setTimes,
    reset: actions.reset,
  }
)(Greet);
