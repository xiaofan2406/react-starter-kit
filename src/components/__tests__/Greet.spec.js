import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { selectors } from 'store';
import { Component, mapStateToProps } from '../Greet';

const props = {
  message: 'World',
  setMessage: jest.fn(),
  times: 1,
  setTimes: jest.fn(),
  reset: jest.fn(),
  greeting: 'World',
};

test('Greet matches snapshot', () => {
  const { container } = render(<Component {...props} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('mapStateToProps', () => {
  const mockState = { greet: {} };
  const getMessageSpy = jest.spyOn(selectors.greet, 'getMessage');
  const getTimesSpy = jest.spyOn(selectors.greet, 'getTimes');
  const getGreetingSpy = jest.spyOn(selectors.greet, 'getGreeting');
  mapStateToProps(mockState);

  expect(getMessageSpy).toHaveBeenLastCalledWith(mockState);
  expect(getTimesSpy).toHaveBeenLastCalledWith(mockState);
  expect(getGreetingSpy).toHaveBeenLastCalledWith(mockState);
});

test('setMessage called on messageInput change', () => {
  const { setMessage } = props;
  const { getByTestId } = render(<Component {...props} />);

  expect(setMessage).toHaveBeenCalledTimes(0);

  const messageInput = getByTestId('messageInput');
  fireEvent.change(messageInput, { target: { value: 'hey' } });

  expect(setMessage).toHaveBeenCalledTimes(1);
  expect(setMessage).toHaveBeenCalledWith('hey');
});

test('setTimes called on messageInput change', () => {
  const { setTimes } = props;
  const { getByTestId } = render(<Component {...props} />);

  expect(setTimes).toHaveBeenCalledTimes(0);

  const timesInput = getByTestId('timesInput');
  fireEvent.change(timesInput, { target: { value: '3' } });

  expect(setTimes).toHaveBeenCalledTimes(1);
  expect(setTimes).toHaveBeenCalledWith(3);
});
