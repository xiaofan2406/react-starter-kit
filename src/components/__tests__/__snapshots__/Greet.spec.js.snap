// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Greet matches snapshot 1`] = `
.emotion-0 {
  margin: 0px auto;
  width: 480px;
}

.emotion-0 > .controls {
  text-align: center;
}

.emotion-0 > .controls input {
  border: none;
  border-bottom: 1px solid #f5f5f5;
  outline: none;
  width: 100px;
  text-align: center;
}

.emotion-0 > .controls input:focus {
  border-bottom: 1px solid #00bcd4;
}

.emotion-0 > .controls > button {
  margin-left: 1em;
  outline: none;
  background: none;
  border: 1px solid #f5f5f5;
}

.emotion-0 > .controls > button:hover {
  border-color: #00bcd4;
  color: #00bcd4;
}

.emotion-0 > .result {
  height: 280px;
  overflow-y: scroll;
  margin-top: 16px;
  font-size: 18px;
  font-family: cursive;
}

.emotion-0 > .result::-webkit-scrollbar {
  width: 10px;
  background-color: rgba(0,0,0,0);
  border-radius: 40px;
}

.emotion-0 > .result::-webkit-scrollbar:hover {
  background-color: rgba(0,0,0,0.09);
}

.emotion-0 > .result::-webkit-scrollbar-thumb:vertical {
  background-color: rgba(0,0,0,0.5);
  border-radius: 40px;
  background-clip: padding-box;
  border: 2px solid rgba(0,0,0,0);
  min-height: 10px;
}

.emotion-0 > .result::-webkit-scrollbar-thumb:vertical:active {
  background-color: rgba(0,0,0,0.61);
}

.emotion-0 > .result:before {
  content: '"';
}

.emotion-0 > .result:after {
  content: '"';
}

<div
  class="emotion-0"
>
  <div
    class="controls"
  >
    <span>
      Say 
    </span>
    <input
      data-testid="messageInput"
      type="text"
      value="World"
    />
     
    <input
      data-testid="timesInput"
      min="1"
      type="number"
      value="1"
    />
     
    times
    <button
      type="button"
    >
      Clear
    </button>
  </div>
  <div
    class="result"
  >
    Hello 
    World
  </div>
</div>
`;
